File: app/src/main/java/com/divercity/app/core/base/BaseFragment.java
package com.divercity.app.core.base;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.divercity.app.core.navigation.Navigator;
import com.divercity.app.di.Injectable;

import javax.inject.Inject;

/**
 * Created by lucas on 27/09/2018.
 */

public abstract class BaseFragment extends Fragment implements Injectable {

    @Inject
    public ViewModelProvider.Factory viewModelFactory;
    @Inject
    public Navigator navigator;

    private ViewDataBinding viewDataBinding;
    private ViewModel viewModel;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        viewDataBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        viewModel = viewModel == null ? getViewModel() : viewModel;
        if (getBindingVariable() != -1)
            viewDataBinding.setVariable(getBindingVariable(), viewModel);
        return viewDataBinding.getRoot();
    }

    public abstract int getLayoutId();

    public abstract ViewModel getViewModel();

    public abstract int getBindingVariable();

    public ViewDataBinding getViewDataBinding() {
        return viewDataBinding;
    }

    public void hideKeyboardInFragment() {
        InputMethodManager imm = (InputMethodManager) getView().getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(getView().getWindowToken(), 0);
    }

}

File: app/src/main/java/com/divercity/app/core/base/BaseFragment.kt
package com.divercity.app.core.base

import android.arch.lifecycle.ViewModelProvider
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import com.divercity.app.core.navigation.Navigator
import com.divercity.app.features.home.HomeActivity
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.activity_layout.*
import javax.inject.Inject

/**
 * Created by lucas on 24/10/2018.
 */

abstract class BaseFragment : Fragment() {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var navigator : Navigator

    abstract fun layoutId(): Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
        inflater.inflate(layoutId(), container, false)

    protected fun showProgress() = progressStatus(View.VISIBLE)

    protected fun hideProgress() = progressStatus(View.GONE)

    private fun progressStatus(viewStatus: Int) =
            with(activity) { if (this is BaseActivity || this is HomeActivity) this.include_loading.visibility = viewStatus }

    fun hideKeyboard() {
        view?.run {
            val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(windowToken, 0)
        }
    }
}